### ISSUE TITLE
Resource Tags

### ISSUE THREAD

**Issue by wiederhold** [byu-oit/handel#332]
Currently Handel is excellent at auto-injecting the following required tags:

  **name, app, env**

Unfortunately the following tags are totally optional:

  **team, data-sensitivity, platform, domain**

While all of these tags are currently required by policy, our mechanisms for enforcement aren't really effective. We can start shutting down non-compliant resources, but identifying who to notify of that action will be especially problematic without even a '**team**' tag.

* Recommend implementing some form of a **NetID** tag that corresponds with the person deploying resources. With that information we can track down the corresponding person most likely to be responsible/knowledgable about the deployment. This would be updated with each deployment.

The following tags remain largely unadopted for a variety of reasons:

  **data-sensitivity, platform, domain**

If we're going to require them by policy, we ought to provide some way to help folks meet that policy. As Handel gains traction as the first option for deployment, we could cover **A LOT** of ground by simply requiring these tags at deployment.

* Recommend implementing some form of deployment switch **-t** for the remaining tags **OR** integrating an interview process in into the **deployPhase switch** @ line 74

If this is what the Steering Committee intended with the 'enforced at deployment', I'm in vociferous support. If the underlying resources don't support tags, no worries. If Handel doesn't support the underlying service, same. With Handel support we'll make **SIGNIFICANT** tactical progress on this issue.

**Comment by ThatJoeMoore** [comment]
Perhaps what we need is an equivalent to Cloudformation's stack tags - some top-level tags in the handelfile that get applied to all resources.

We wouldn't want to start enforcing strict BYUisms in Handel core, as it's been purposefully kept very generic, but perhaps we could add a list of required tags to the account config files and have Handel validate that they've been set? Then we get this kind of enforcement in a generic way.

**Comment by wiederhold** [comment]
Sounds like an excellent strategy.

I totally get the open-source, share with the community thing--as long as the University's needs are top of mind as well.

Ultimately we could benefit from any form of personnel identification. BYUisms could be enforced post hoc with that information. And, that IS a generalizable issue. From a maintenance perspective, every CI in any medium to large size size organization requires an identifiable owner, a point of contact.

Currently we have too many resources deployed where our only clue about the stack are inferences from an app name or the id of the account it's deployed in. That is not sustainable.

**Comment by ThatJoeMoore** [comment]
Because of the way we intend for Handel  (ie, with Handel-codepipeline) to be used, having a tag identifying the user who deployed a resource isn't really possible, as it'll almost always be a pipeline. In addition, we probably don't want to go around causing updates on every resource every time there's a deployment from a different person, so tying each resource to the person who deployed it or caused it to be deployed might not be the best approach. If we enforce a rule of requiring the team tag, wouldn't that get us 95% of the way to tracking down an owner?

I would like to see us have a way to track a handelfile back to the pipeline that deployed it, probably using stack-level tags on the CFN stacks.

Perhaps we could also autotag resources coming out if a pipeline with the repository that's feeding into it? Then, we could see who's committing. I'm wary of using up too many tags, but the sort-of-recent increases to the tag limits make me less concerned.

Keeping track of things like "Person X pushed a commit through pipeline P which caused resource R to be created" would require us to add some kind of remote state storage into Handel, which is something that has encountered resistance thus far because it would dramatically increase Handel's complexity. I've seen a few use cases for it, though, so maybe we need to start thinking about how it would look.

**Comment by ndpete** [comment]
I have been doing some tagging today I think a global or top level tag definition in handel would be convenient. That way the same set of tags could get applied to all the resources that support it. I find I'm copying and pasting the same tag block to each service that supports tags.

**Comment by wiederhold** [comment]
☝️  cough, cough

**Comment by ThatJoeMoore** [comment]
Okay, so #362 handles the application-level tagging thing.  So, I think we really have four things left:

- [x] ~~application-level tagging~~
- [ ] Allow specifying extra tags with the CLI
- [ ] Use ☝️ to auto-tag stacks coming from Handel-codepipeline with info about the pipeline
- [ ] Add ability to enforce required tags using account config file
- [ ] Tag Handel account-wide resources ( #360 )

With a potential future awesome bonus feature:

- [ ] allow an account config file to specify a lambda or something that validates and enforces organization tagging files.

@dsw88 @wiederhold @ndpete do you think that would get us to a good place? I'm hoping we can balance the needs of BYU, flexibility inside of Handel, and ease of use by doing these things.

**Comment by ndpete** [comment]
I think those sound pretty reasonable.
Tagging stacks coming out of a pipeline would be amazing to get better insight into where resources are coming from.
I really like the idea of being able to enforce tags as people try to deploy pipelines and handle resources.

**Comment by bwinterton** [comment]
I actually really like the concept of being able to tag the resources with information about the pipeline. That could be useful to track back where certain resources came from. 

I also like the idea of enforcing required tags in the account config file. Handel could just refuse to deploy if the account config specifies a tag is required that wasn't set in the handel file. That can be done in a very generic/non-BYU specific way and could be super useful for us going forward.

**Comment by dsw88** [comment]
I also really like injecting tags at run-time via CLI parameters! I also believe that implementing enforcement of tags from the account config file is important, as well as tagging the resources Handel creates to support service provisioners.

As for the Lambda to validate organizational tagging rules, we at BYU already have some AWS Config rules that do this, although nobody has really used them for much thus far. As for a generic feature for Handel users, I think this is a good idea, although I wonder whether it should be a separate project or plugin from the main project.

**Comment by ThatJoeMoore** [comment]
For the pie-in-the-sky validation, I was thinking something as simple as having a `stack_validation_function` in the account config. When Handel runs, it makes a synchronous call to that lambda, passing it the Handelfile data structure. Then, the lambda, which is written by the organization, can do whatever validation it wants to on the stack. It could, for example, go beyond tagging and prevent the deployment of a public S3 bucket or something. However, it does sound like AWS Config might be a better fit. I'm still not super familiar with it; can it do dynamic things like 'pull in the list of teams from Service Now and make sure that the `team` tag is one of them'?

**Comment by bwinterton** [comment]
I was having a similar thought to what @ThatJoeMoore is saying. Basically just specifying a lambda to validate and documenting the message format that would be sent to the lambda. That way it is generic and easy to implement for any organization if they want some level of validation. 

Config rules can definitely pull in a list of teams and then validate the `team` tag. They are essential just lambdas that get triggered when the account has something change. Config rules also work in a post deployment sense. So I think the validation lambda makes sense to prevent bad deployments, but we could also add config rule support to Handel to deploy other config rules.

**Comment by ThatJoeMoore** [comment]
#364, adding support for listing required tags in the account config, is ready for review.

**Comment by wiederhold** [comment]
> I also really like injecting tags at run-time via CLI parameters! I also believe that implementing enforcement of tags from the account config file is important, as well as tagging the resources Handel creates to support service provisioners.

As for the Lambda to validate organizational tagging rules, we at BYU already have some AWS Config rules that do this, although nobody has really used them for much thus far. As for a generic feature for Handel users, I think this is a good idea, although I wonder whether it should be a separate project or plugin from the main project.

☝️ I would strongly vote for including it as a plugin for the main project. As @ThatJoeMoore and @bwinterton have already suggested, config standardization is likely to be a high priority for any organization. This is am amazing value-add for Handel at BYU, AND this is WAY better than AWS Config in its present state. More to the point, this gets config rules closer to the dev pipeline than any other mechanism anyone's suggested. HUGE thumbs up... this is IMO brilliant!

**Comment by ThatJoeMoore** [comment]
I've now got PRs in for all of the simple changes listed above.

Now, we'll need to discuss a plan to apply all of these changes, which will include redeploying pipelines, manually tagging some resources, and distributing new account config files.

**Comment by dsw88** [comment]
I reviewed and merged the PRs, thanks for implementing them! 

We can pretty easily update the account config files. Redeploying pipelines will be a little difficult because it's complicated by the fact that some people still haven't moved to the Main VPC yet, so we can't just auto-update everyone to use the latest account config file quite yet. 

That incidentally is the main reason I chose to embed the account config file in the pipeline rather than put it in a shared location, to allow for people using multiple potentially conflicting versions of it across different pipelines.
